lib/SocketConnection.cs

```diff
@@ -11,49 +11,76 @@ class SocketConnection
         public readonly string Url;
         public event EventHandler<string> OnData;
         private ClientWebSocket Socket;
+        private int maxRetryCount = 4;
+        private int secondsBetweenRetry = 2;
 
         public SocketConnection(string url)
         {
             this.Url = url;
             Connect();
         }
 
-        public async void SendData(ArraySegment<byte> data){
+        public async void SendData(ArraySegment<byte> data)
+        {
             await Socket.SendAsync(data, WebSocketMessageType.Text, true, CancellationToken.None);
         }
 
         async void Connect()
         {
-            try
+            await TryConnect();
+        }
+
+        private async System.Threading.Tasks.Task TryConnect()
+        {
+            int retryCounter = 0;
+            while (retryCounter < maxRetryCount)
             {
-                Socket = new System.Net.WebSockets.ClientWebSocket();
-                await Socket.ConnectAsync(new Uri(this.Url), CancellationToken.None);
+                try
+                {
+                    Socket = new System.Net.WebSockets.ClientWebSocket();
+                    await Socket.ConnectAsync(new Uri(this.Url), CancellationToken.None);
+
+                    var receiveBytes = new byte[4096];
+                    var receiveBuffer = new ArraySegment<byte>(receiveBytes);
 
-                var receiveBytes = new byte[4096];
-                var receiveBuffer = new ArraySegment<byte>(receiveBytes);
+                    while (Socket.State == WebSocketState.Open)
+                    {
+                        var receivedMessage = await Socket.ReceiveAsync(receiveBuffer, CancellationToken.None);
+                        if (receivedMessage.MessageType == WebSocketMessageType.Close)
+                        {
+                            await
+                                Socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "closing websocket",
+                                    CancellationToken.None);
+                        }
+                        else
+                        {
+                            var messageBytes = receiveBuffer.Skip(receiveBuffer.Offset).Take(receivedMessage.Count).ToArray();
 
-                while (Socket.State == WebSocketState.Open)
+                            var rawMessage = new UTF8Encoding().GetString(messageBytes);
+                            OnData.Invoke(this, rawMessage);
+                        }
+                    }
+                    break;
+                }
+                catch (Exception e)
                 {
-                    var receivedMessage = await Socket.ReceiveAsync(receiveBuffer, CancellationToken.None);
-                    if (receivedMessage.MessageType == WebSocketMessageType.Close)
+                    int sleepTimeSeconds = Convert.ToInt32(Math.Pow(secondsBetweenRetry, retryCounter + 1));
```

Comment:
Has an exponential backoff which felt pretty slow TBH, might want to consider tweaking with a linear time.